consensus/safety-rules/src/safety_rules.rs

```diff
@@ -122,22 +123,33 @@ impl<T: Payload> TSafetyRules<T> for SafetyRules<T> {
     /// @TODO verify QC correctness
     /// @TODO verify epoch on vote proposal
     fn construct_and_sign_vote(&mut self, vote_proposal: &VoteProposal<T>) -> Result<Vote, Error> {
+        debug!("Incoming vote proposal to sign.");
         let proposed_block = vote_proposal.block();
 
-        if proposed_block.round() <= self.persistent_storage.last_voted_round()? {
+        let last_voted_round = self.persistent_storage.last_voted_round()?;
+        if proposed_block.round() <= last_voted_round {
+            debug!(
+                "Vote proposal is old {} < {}",
+                proposed_block.round(),
+                last_voted_round
+            );
             return Err(Error::OldProposal {
                 proposal_round: proposed_block.round(),
                 last_voted_round: self.persistent_storage.last_voted_round()?,
             });
         }
 
-        let respects_preferred_block = proposed_block.quorum_cert().certified_block().round()
-            >= self.persistent_storage.preferred_round()?;
+        let preferred_round = self.persistent_storage.preferred_round()?;
+        let respects_preferred_block =
+            proposed_block.quorum_cert().certified_block().round() >= preferred_round;
 
         if !respects_preferred_block {
```

Comment:
nit: any reason why you don't just do the comparison directly? e.g.,

if proposed_block.quorum_cert().certified_block().round() < preferred_round {
...
}

The ! in front of the greater than or equals to makes it non-obvious to me.

Also, there's probably a reason for this, but when logging the failure, we log proposed_block.round(), rather than proposed_block.quorum_cert().certified_block().round()?